packages/yarnpkg-pnp/sources/esm-loader/loader.ts

```diff
@@ -0,0 +1,192 @@
+import {NativePath, PortablePath}          from '@yarnpkg/fslib';
+import {parse, init}                       from 'cjs-module-lexer';
+import fs                                  from 'fs';
+import moduleExports                       from 'module';
+import path                                from 'path';
+import {fileURLToPath, pathToFileURL, URL} from 'url';
+
+import * as nodeUtils                      from '../loader/nodeUtils';
+import {PnpApi}                            from '../types';
+
+function isValidURL(str: string) {
+  try {
+    new URL(str);
+    return true;
+  } catch {
+    return false;
+  }
+}
+
+const builtins = new Set([...moduleExports.builtinModules]);
+
+const pathRegExp = /^(?![a-zA-Z]:[\\/]|\\\\|\.{0,2}(?:\/|$))((?:node:)?(?:@[^/]+\/)?[^/]+)\/*(.*|)$/;
+
+async function exists(path: string) {
+  try {
+    await fs.promises.access(path, fs.constants.R_OK);
+    return true;
+  } catch { }
+  return false;
+}
+
+export async function resolve(
+  originalSpecifier: string,
+  context: any,
+  defaultResolver: any
+) {
+  const {findPnpApi} = (moduleExports as unknown as { findPnpApi?: (path: NativePath) => null | PnpApi });
```

Comment:
How is `findPnpApi` injected in the loader? Is that done by the `.pnp.cjs` file?

Reply:
Yes, `.pnp.cjs` runs before the loader is executed